incubator/virtualcluster/cmd/kubectl-vc/create.go

```diff
@@ -48,64 +48,87 @@ const (
 	pollStsTimeoutSec = 120
 )
 
-var createCommand = cli.Command{
-	Name:  "create",
-	Usage: "Create a new VirtualCluster",
-	Flags: []cli.Flag{
-		cli.StringFlag{
-			Name:  "filename, f",
-			Usage: "the configuration to apply",
-		},
-		cli.StringFlag{
-			Name:  "output, o",
-			Usage: "path to the kubeconfig that is used to access virtual cluster",
+type CreateOptions struct {
+	client     client.Client
+	vcclient   vcclient.Interface
+	fileName   string
+	outputPath string
+}
+
+func NewCmdCreate(f Factory) *cobra.Command {
+	o := &CreateOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "create",
+		Short: "Create a new VirtualCluster",
+		Run: func(cmd *cobra.Command, args []string) {
+			CheckErr(o.Complete(f, cmd))
+			CheckErr(o.Validate(cmd))
+			CheckErr(o.Run())
 		},
-	},
-	Action: func(cctx *cli.Context) error {
-		filename := cctx.String("filename")
-		if filename == "" {
-			return errors.New("must specific --filename,-f flag")
-		}
+	}
 
-		outputPath := cctx.String("output")
-		if outputPath == "" {
-			return errors.New("must specific --output,-o flag")
-		}
+	cmd.Flags().StringVarP(&o.fileName, "filename", "f", "", "the configuration to apply")
+	cmd.Flags().StringVarP(&o.outputPath, "output", "o", "", "path to the kubeconfig that is used to access virtual cluster")
 
-		vccli, err := newVCClient()
-		if err != nil {
-			return err
-		}
+	return cmd
+}
 
-		fileBytes, err := getYamlContent(filename)
-		if err != nil {
-			return errors.Wrapf(err, "read \"%s\"", filename)
-		}
+func (o *CreateOptions) Complete(f Factory, cmd *cobra.Command) error {
+	var err error
+	o.vcclient, err = f.VirtualClusterClientSet()
+	if err != nil {
+		return err
+	}
 
-		vc := &tenancyv1alpha1.VirtualCluster{}
-		if err = yaml.Unmarshal(fileBytes, vc); err != nil {
-			return err
-		}
+	o.client, err = f.GenericClient()
+	if err != nil {
+		return err
+	}
 
-		kubecfgBytes, err := createVirtualCluster(vccli, vc)
-		if err != nil {
-			return err
-		}
+	return nil
+}
 
-		// write tenant kubeconfig to outputPath.
-		return ioutil.WriteFile(outputPath, kubecfgBytes, 0644)
-	},
+func (o *CreateOptions) Validate(cmd *cobra.Command) error {
+	if len(o.fileName) == 0 {
+		return UsageErrorf(cmd, "--filename,-f should not be empty")
+	}
+	if len(o.outputPath) == 0 {
+		return UsageErrorf(cmd, "--output,-o should not be empty")
+	}
+	return nil
 }
 
-func createVirtualCluster(vccli vcclient.Interface, vc *tenancyv1alpha1.VirtualCluster) ([]byte, error) {
-	cv, err := vccli.TenancyV1alpha1().ClusterVersions().Get(vc.Spec.ClusterVersionName, metav1.GetOptions{})
+func (o *CreateOptions) Run() error {
+	fileBytes, err := getYamlContent(o.fileName)
 	if err != nil {
-		return nil, errors.Wrapf(err, "required cluster version not found")
+		return errors.Wrapf(err, "read \"%s\"", o.fileName)
 	}
 
-	apiSvcPort, err := getAPISvcPort(cv.Spec.APIServer.Service)
+	vc := &tenancyv1alpha1.VirtualCluster{}
+	if err = yaml.Unmarshal(fileBytes, vc); err != nil {
```

Comment:
Do you want to use the apimachinery scheme to decode this for you instead of having to use the yaml parsers. All you have to do is use the scheme to add the tenancy scheme then call `DecodeInto` and pass a runtime.Object and it will parse for you.

```
	content, err := ioutil.ReadFile(o.fileName)
	// handle err
	codecs := serializer.NewCodecFactory(scheme)
	err := runtime.DecodeInto(codecs.UniversalDecoder(), content, vc)
        // handle err
```

for a full example I'm doing this here - https://github.com/kubernetes-sigs/controller-runtime/blob/e6c724240cc44ff8cb96ffeae019341f6d44f24d/pkg/config/config.go#L99-L120

Reply:
can this codec recognize the json or yaml format?

Reply:
oh i see, json works fine as well